src/platform_impl/android/mod.rs

```diff
@@ -353,128 +354,143 @@ impl<T: 'static> EventLoop<T> {
             trace!("No main event to handle");
         }
 
-        // Process input events
-        self.android_app.input_events(|event| {
-            let mut input_status = InputStatus::Handled;
-            match event {
-                InputEvent::MotionEvent(motion_event) => {
-                    let window_id = window::WindowId(WindowId);
-                    let device_id = event::DeviceId(DeviceId);
+        // temporarily decouple `android_app` from `self` so we aren't holding
+        // a borrow of `self` while iterating
+        let android_app = self.android_app.clone();
 
-                    let phase = match motion_event.action() {
-                        MotionAction::Down | MotionAction::PointerDown => {
-                            Some(event::TouchPhase::Started)
-                        }
-                        MotionAction::Up | MotionAction::PointerUp => {
-                            Some(event::TouchPhase::Ended)
-                        }
-                        MotionAction::Move => Some(event::TouchPhase::Moved),
-                        MotionAction::Cancel => {
-                            Some(event::TouchPhase::Cancelled)
-                        }
-                        _ => {
-                            None // TODO mouse events
-                        }
-                    };
-                    if let Some(phase) = phase {
-                        let pointers: Box<
-                            dyn Iterator<Item = android_activity::input::Pointer<'_>>,
-                        > = match phase {
-                            event::TouchPhase::Started
-                            | event::TouchPhase::Ended => {
-                                Box::new(
-                                    std::iter::once(motion_event.pointer_at_index(
-                                        motion_event.pointer_index(),
-                                    ))
-                                )
-                            },
-                            event::TouchPhase::Moved
-                            | event::TouchPhase::Cancelled => {
-                                Box::new(motion_event.pointers())
+        // Process input events
+        match android_app.input_events_iter() {
+            Ok(mut input_iter) => {
+                loop {
+                    let read_event = input_iter.next(|event| {
```

Comment:
You can merge all of that by mapping an iterator `map` and then doing `for event in <your_iterator>`.

Reply:
The input event iterator isn't a standard Rust Iterator because it's a lending iterator which is largely why the ergonomics are a bit awkward here.

I've factored out a `handle_input_event` function though and that helps a lot with making this readable.